src/AIAPI/MainGame/GameCustomFunctionController.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+// using ActionGame;
+using ExtensibleSaveFormat;
+using KKAPI.Maker;
+using UnityEngine;
+using Manager;
+using AIProject;
+using AIChara;
+using CharaCustom;
+
+namespace KKAPI.MainGame
+{
+    /// <summary>
+    /// Base type for custom game extensions.
+    /// It provides many useful methods that abstract away the nasty hooks needed to figure out when the state of the game
+    /// changes.
+    /// 
+    /// This controller is a MonoBehaviour that is created upon registration in <see cref="GameAPI.RegisterExtraBehaviour{T}"/>.
+    /// The controller is created only once. If it's created too late it might miss some events.
+    /// It's recommended to register controllers in your Start method.
+    /// </summary>
+    public abstract class GameCustomFunctionController : MonoBehaviour
+    {
+        private static EnvironmentSimulator _cycle;
```

Comment:
I'm trying to keep the original property/field names when changing types, to make it easier to see what was converted.

Reply:
The only members that matter are the public ones, the guts could be completely replaced, so you could look at it in terms of which public members got implemented or not. It depends on how you want to go about porting it though.